File: countingSort/countingSort.py
# Counting Sort in python
# Optimized for both positive and negative integers
# works only ascending order
def countingSort(array):
    min  = min(array)
    max = max(array)
    
    temp = {key : 0 for key in range(min, max)}
    for value in array:
        temp[value] += 1
    
    array = list()
    for key in range(min, max):
        while temp[key] > 0:
            array.append(key)
            temp[key] -= 1
    return

File: countingSort/countingSort.cpp
// Counting Sort in python
// Optimized for both positive and negative integers
// works only in Ascending order

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void countingSort(vector<int> &arr){
    int min = *min_element(arr.begin(), arr.end());
    int max = *max_element(arr.begin(), arr.end());

    vector<int> temp(max -min +1);
    for(int i = 0; i < (int)temp.size(); ++i)   temp[i] = 0;
    for(int i = 0; i < (int)arr.size(); ++i)    ++temp[arr[i] -min];
    for(int i = 0, j = 0; i < (int)temp.size(); ++i){
        while(temp[i] > 0){
            arr[j] = i +min;
            ++j;
            --temp[i];
        }
    }
}